### PR TITLE
Use more fine-grained locks in Peers

### PR DESCRIPTION
**Pull request by ljedrz**
Isolated from https://github.com/AleoHQ/snarkOS/pull/1318 for more convenient reviewing and testing; I seem to be bleeding peers (due to them `spamming unconfirmed blocks (frequency = 5)`) after some time, but this could be caused by some of the network protocol rules that might need to be relaxed (as these changes enable faster communication, which might have been unexpected before).

### CODE REVIEW

**Comment by ljedrz** [comment]
note: the second commit from https://github.com/AleoHQ/snarkOS/pull/1318 applies cleanly on top of this PR (which is its first commit)

**Comment by howardwu** [comment]
More of a high-level convention question - when should we be using `Arc<RwLock<Foo>>` as opposed to `RwLock<Foo>`? Want to make sure we maintain consistency in design choice.

File: src/network/server.rs
Comment (niklaslong) [inline src/network/server.rs]: 👍

File: src/network/peers.rs
Comment (niklaslong) [inline src/network/peers.rs]: Note: we might want structs in place of the tuples here (and further up). This is more of a design decision but some unification may not be a bad thing for all these collections.

**Comment by ljedrz** [comment]
>when should we be using `Arc<RwLock<Foo>>` as opposed to `RwLock<Foo>`? Want to make sure we maintain consistency in design choice

Whenever there's an object that requires shared ownership, an `Arc` provides the ability to shallowly clone objects (and pass them on to tasks, threads etc.) that should maintain a consistent state (as opposed to e.g. some `String` that can just be cloned around and it doesn't need to remain consistent) or objects that are just too heavy to clone deeply (even if their state doesn't need to remain consistent everywhere), like large collections.

Since `Arc` use does incur a little bit of runtime overhead, it is also better if multiple objects can be under a single `Arc` (e.g. having `Arc<Peers>` as opposed to plain `Peers` with its children all being `Arc`ed), as long as they can remain immutable or have a lock over them (since `Arc` requires immutability).

File: src/network/peers.rs
Comment (niklaslong) [inline src/network/peers.rs]: We should probably store these as `SocketAddr`.

File: src/network/peers.rs
Comment (niklaslong) [inline src/network/peers.rs]: Do we want a `Nonce` type at some point?

**Review by niklaslong — APPROVED** [review]
LGTM, left a few comments for consideration 👍

**Comment by ljedrz** [comment]
@niklaslong agreed with your points, will propose them in a future PR in order to keep this PR self-contained.

**Comment by ljedrz** [comment]
@howardwu this PR can cause RPC to hang after it's been running for a while; I need to find the cause before it can be merged (unless this has already been observed with the main branch)

**Comment by ljedrz** [comment]
I think I was able to find the issue, but I'll keep testing for a bit longer before un-drafting.

**Comment by ljedrz** [comment]
Yep, it appears that the fix did the trick :+1:; I can no longer reproduce the RPC hang.

**Pull request merged**
